---
title: "Assignment 1"
author: "Casey Hon"
output:
  html_document:
    df_print: paged
---

# Introduction

Individuals with HIV are at higher risk to have tuberculosis (TB), and this is often due to infection with *Mycobacterium tuberculosis* (*M. tuberculosis*) that rapidly progresses to disease (Correa-Macedo et al., 2021). Alveolar macrophages (AMs) are the first cells in the immune system that interact with *M. tuberculosis*, however their interaction with HIV and antiretroviral therapy (ART) is still unknown (Correa-Macedo et al., 2021). Thus, this study aimed to investigate the transcriptomic and epigenetic response of AMs to *M. tuberculosis* and how HIV and ART play a part in these mechanisms.

AMs were obtained from 16 control subjects who were HIV-free (HC), 20 persons living with HIV receiving ART (PLWH), and 14 subjects who received ART as preexposure prophylaxis (PrEP) to prevent HIV infection. Each sample was challenged with *M. tuberculosis* in vitro. This dataset was of interest to me because of my lack of knowledge about HIV, and recognizing how there are social stigmas around this condition. Through this study, I learned that TB is actually the leading cause of death from a single pathogen, with 1.4 million deaths in 2019 alone. This included 208,000 people who had HIV, and people with HIV actually had a 2.0-fold higher mortality. This interested me that even given the severity, we are still unsure as to how HIV and it's related therapy causes this.

Therefore, this notebook aims to analyze the RNA-seq data from this study on the gene level.

## Downloading the Data

This dataset is from the study *Alveolar macrophages from persons living with HIV show impaired epigenetic response to Mycobacterium tuberculosis* (Correa-Macedo et al., 2021).

```{r, message = FALSE}
suppressWarnings({
  if (! requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
  }
  if (! requireNamespace("Biobase", quietly = TRUE)) {
    BiocManager::install("Biobase")
  }
  if (! requireNamespace("GEOquery", quietly = TRUE)) {
    BiocManager::install("GEOquery")
  }
  if (! requireNamespace("edgeR", quietly = TRUE)) {
    BiocManager::install("edgeR")
  }
  if (! requireNamespace("biomaRt", quietly = TRUE)) {
    BiocManager::install("biomaRt")
  }
  if (! requireNamespace("ggplot2", quietly = TRUE)) {
    BiocManager::install("ggplot2")
  }
  if (! requireNamespace("affy", quietly = TRUE)) {
    BiocManager::install("affy")
  }
  if (! requireNamespace("PoiClaClu", quietly = TRUE)) {
    BiocManager::install("PoiClaClu")
  }
})

# load non-normalized counts
dataset_name <- "GSE165708"
GSE165708 <- GEOquery::getGEO(dataset_name, GSEMatrix = FALSE, getGPL= FALSE)
if (!dir.exists(paste0("./", dataset_name))) {
  supp_files <- GEOquery::getGEOcountsFiles(dataset_name)
  file_names <- rownames(supp_files) 
  counts <- read.delim(file_names[2], header = TRUE, check.names = FALSE)
} else {
  counts <- read.delim("./GSE165708/GSE165708_non-normalized_estimated_counts_matrix.txt.gz")
}
```

**Data information**

```{r}
platform <- names(GEOquery::GPLList(GSE165708))[1]
platform_metadata <- GEOquery::Meta(GEOquery::getGEO(platform))
```

**Platform:** `r platform_metadata$title`

**Organisms:** `r platform_metadata$organism` (taxid: `r platform_metadata$taxid`)

## Pre-cleaning Statistics

Unique gene count: `r length(unique(counts$gene_id))`

```{r group_counts}
## ---- groups_chunk
# split ensembl ID and gene name, code referred to Lecture 4 slides
gene_names <- data.frame(lapply(counts$gene_id, FUN = function(x){unlist(strsplit(x, split = "_"))}))
counts$gene_name <- t(gene_names[2,])
counts$ensembl_id <- t(gene_names[1,])
rownames(counts) <- counts$ensembl_id
groups <- data.frame(lapply(colnames(counts)[2:101], FUN = function(x){unlist(strsplit(x, 
                                                                                        split = "\\."))}))
colnames(groups) <- colnames(counts[2:101])
rownames(groups) <- c("sample_group", "patient", "disease", "challenged") 
# challenged = boolean value for whether or not the sample was challenged with M. tuberculosis, with 0 means non-challenged, 1 means challenged.
groups <- as.data.frame(t(groups))
saveRDS(groups, file = "groups.rds")
group_counts <- data.frame(HC = length(which(groups$sample_group == "HC")), 
                            PrEP = length(which(groups$sample_group == "PrEP")), 
                            PLWH = length(which(groups$sample_group == "PLWH")))
knitr::kable(group_counts, caption = "Number of groups in each condition")
```

Total sample groups: `r unique(groups$sample_group)`

-   **HC** = Control subjects who were HIV-free

-   **PLWH** = Persons living with HIV receiving ART

-   **PrEP** = Subjects who received ART as preexposure prophylaxis

Total Patients: `r length(unique(groups$patient))`

Number of groups challenged: `r length(groups$challenged == 1)`

```{r}
# (code from Lecture 4 slides)
gene_name_counts <- table(counts$gene_name)
knitr::kable(gene_name_counts[which(gene_name_counts > 1)], caption = "Genes with counts greater than 1", col.names = c("Gene", "Frequency"))
```

The only gene counts duplicated are due to similarity between two genes, but they are actually just a variant. For example, upon manually checking the two Ensembl IDs that are both coded as AHRR, one maps to AHRR and the other maps to PDCD6-AHRR, which is the programmed cell death 6 version of the gene. This just reinforces the importance of mapping to the HUGO gene symbols too to further differentiate some of the genes that look like duplicates but actually differ.

Filter out genes that have low counts:

```{r}
# code from Lecture 4
cpms <- edgeR::cpm(counts[,2:101])
rownames(cpms) <- counts$gene_id
# n is set to 28, which is the size of the smallest group of replicates (PrEP)
filtered <- counts[rowSums(cpms > 1) >= 28, ]
counts_filtered <- table(filtered$gene_name)
```

```{r}
knitr::kable(counts_filtered[which(counts_filtered > 1)], col.names = c("Gene", "Frequency"))
```

I used the edgeR protocol recommendations, which states to remove features that has at least 1 read per million in n of the samples, where n is the size of the smallest group of replicates. As we saw in an above table, the smallest group was the PrEP condition with 28 samples. Filtering leaves 13,780 genes, much lower than the original 60,000+. Specifically, `r nrow(counts) - nrow(counts_filtered)` outliers were removed. It also reduces the number of duplicates to just a few that is easier to handle.

## Mapping to HUGO gene symbols

```{r}
# based on code from Lecture 4
mart <- biomaRt::useMart(biomart = "ensembl")
ensembl <- biomaRt::useDataset(dataset = "hsapiens_gene_ensembl", 
                               mart = mart)
filters <- biomaRt::listFilters(mart = ensembl)

converted <- "id_conv.rds"
if (file.exists(converted)) { # check if file already exists before converting again
  id_conv <- readRDS("id_conv.rds")
} else {
  id_conv <- biomaRt::getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                                      filters = c("ensembl_gene_id"),
                                      values = filtered$ensembl_id,
                                      mart = ensembl)
  saveRDS(id_conv, converted)
}
hgnc <- table(id_conv$hgnc_symbol)
colnames(id_conv) <- c("ensembl_id", "hgnc_symbol")
filtered <- merge(filtered, id_conv, by = "ensembl_id")
knitr::kable(hgnc[which(hgnc > 1)], col.names = c("HGNC Symbol", "Frequency"))
```

There were `r sum(id_conv$hgnc_symbol == "")` genes that could not be mapped to HGNC symbols, which is `r sum(id_conv$hgnc_symbol == "")/length(id_conv$hgnc_symbol) * 100`% of all genes. However, gene names were included in the original dataset that are mostly the same as the HGNC symbols I mapped, with `r sum(filtered$gene_name != filtered$hgnc_symbol) / nrow(filtered) * 100` being different (mostly attributed to genes that could not be mapped to any HGNC symbol). I will merge these instead to deal with missing HGNC identifiers. We also see that mapping to HGNC symbols only leaves 4 duplicated genes.

The duplicated genes are: `r names(hgnc[which(hgnc > 1)])`, ignoring the first empty string which represents genes that could not be mapped to HGNC symbols. Since there are only 4 and each one is only duplicated once (8 genes in total), I will just remove all of them.

```{r}
# merge old mapping
filtered$hgnc_symbol <- ifelse(filtered$hgnc_symbol == "", filtered$gene_name, filtered$hgnc_symbol)

# handling duplicated genes manually, since there are only 4
dups <- names(hgnc[which(hgnc > 1)])
for (i in 2:length(dups)) { # ignore first one which is empty hgnc symbols, which we have merged
  filtered <- filtered[-c(which(filtered$hgnc_symbol == dups[i]))]
}
```

Removing duplicates leaves us with `r nrow(filtered)` genes.

## Distribution of Data before Normalization

```{r}
# code from lecture 4
filtered_log_cpm <- log2(edgeR::cpm(filtered[,3:102]))
par(mar = c(6, 4, 4, 4)) # increase bottom margin since x-labels are so long
suppressWarnings(boxplot(filtered_log_cpm, xlab = "", 
        ylab = "log2(CPM) values", las = 2,
        cex = 0.5, cex.lab = 0.75, cex.axis = 0.5, 
        main = "Log2 CPMs of HIV groups"))
mtext("Conditions", side = 1, line = 4.7) # adds x-axis label
```

There were a lot of warnings that Outliers with value -Inf were not drawn. These likely represents certain samples having 0 cpm, since log2(0) gives the -Inf value. We can ignore these since we already filtered these out. There is also the option to remove these outliers from the graph by adding the Outline = FALSE parameter, but I decided against this as it would obscure how the pre-normalized data looks. Also, it's important to note that the x-axis only shows the samples that end with ".0", which represent non-challenged samples. However, the labelling just kept every other sample, but this does indeed include challenged samples as well. This was double checked by rotating the label names to horizontal, which does show the samples that end with ".1".

Based on this box plot, some of the groups have differing log2(cpm) values, which shows the need for normalization.

```{r}
# based on https://mkempenaar.github.io/gene_expression_analysis/chapter-3.html
cols <- scales::hue_pal()(ncol(filtered_log_cpm))
affy::plotDensity(filtered_log_cpm, col = cols,
                  lty = c(1:ncol(filtered_log_cpm)), 
                  main = "Pre-normalization counts distribution")
legend("topright", colnames(filtered_log_cpm), 
       lty = c(1:ncol(filtered_log_cpm)),
       col = cols,
       cex = 0.6)
```

Based on this density plot, we can see that most of the groups follow a very similar distribution, but none of them follow a normal distribution.

## Data Normalization

The Trimmed Mean of M-values (TMM) method of normalization will be used as it is a technique for RNA-seq data. \*\* TO-DO: elaborate \*\*

```{r filt_chunk, cache.vars=filtered_mat}
#code from lecture 4
filtered_mat <- as.matrix(filtered[, 3:102])
saveRDS(filtered_mat, file = "filtered_mat.rds")
rownames(filtered_mat) <- filtered$ensembl_id
dge <- edgeR::DGEList(counts = filtered_mat, 
                      group = groups$sample_group)
dge <- edgeR::calcNormFactors(dge)
normalized_counts <- edgeR::cpm(dge)
normed_log_counts <- log2(normalized_counts)
saveRDS(normed_log_counts, "normed_log_counts.rds")
```

We now have a matrix object that has 100 columns representing the 100 samples, 13,757 rows representing the genes, and the Ensembl Gene IDs are the rownames. The values are the logged, normalized counts.

```{r}
# plot normalized counts
par(mar = c(6, 4, 4, 4)) 
suppressWarnings(boxplot(normed_log_counts, xlab = "", 
        ylab = "log2(CPM) values", las = 2,
        cex = 0.5, cex.lab = 0.75, cex.axis = 0.5, 
        main = "Log2 CPMs of HIV groups after normalization"))
mtext("Conditions", side = 1, line = 4.7) # adds x-axis label
```

Based on the boxplot, the normalized counts have log2(cpm) values that are much more similar.

```{r}
cols <- scales::hue_pal()(ncol(normed_log_counts))
affy::plotDensity(normed_log_counts, col = cols,
                  lty = c(1:ncol(normed_log_counts)), 
                  main = "Normalized counts distribution")
legend("topright", colnames(normed_log_counts), 
       lty = c(1:ncol(normed_log_counts)),
       col = cols,
       cex = 0.6)
```

The density plot is very slightly changed after normalization, but still shows a bit more similarity in distribution across groups. This shows that the data did not have a lot of technical variation to begin with, but normalizing it just in case is still important.

We can also calculate a distance metric for each combination of samples to check for variation within the sample groups. I will perform Multi-Dimensional Scaling (MDS) to display distances in a 2D-plot. MDS uses Poisson Distance as it's distance metric. Since there are three sample groups, we should expect to see three separated clusters.

```{r plot_chunk}
## ---- plot_chunk
# MDS plot to inspect sample separation
# based on https://mkempenaar.github.io/gene_expression_analysis/ and https://www.bioconductor.org/help/course-materials/2014/SeattleOct2014/B02.1.1_RNASeqLab.html
# PoissonDistance function takes original count matrix, un-normalized, with samples as rows
pois_distances <- PoiClaClu::PoissonDistance(t(filtered_mat))
pois_dist_mat <- as.matrix(pois_distances$dd)
mds <- data.frame(cmdscale(pois_dist_mat))
mds <- cbind(mds, groups)
saveRDS(mds, file = "mds.rds")
ggplot2::qplot(X1, X2, color = sample_group, data = mds, main = "MDS plot of Poisson Distances by sample group")
```

Based on the MDS plot, we do not really see clear clustering by sample group. There is a pattern that the Healthy Control (HC) group (in pink) tends to be on the left side, the Persons living with HIV (PLWH) group (in green) tending to be more on the right, and the pre-exposure prophylaxis (PrEP) group (in blue) towards the bottom middle. However, compared to other clearly-clustered examples, this is definitely a very obscure distinction between sample groups. I also tried plotting by the other conditions, by patient or by challenged status, but the clustering was even more obscure.

## Results

```{r}
rownames(normed_log_counts) <- filtered$hgnc_symbol
final_result <- as.data.frame(normed_log_counts)
```

The final result is a dataframe with `r ncol(final_result)` numeric columns that represent the 100 samples of this experiment. The rownames are the unique HUGO symbols. The colnames are the original sample names from the experiment.

## References

Correa-Macedo, Wilian et al. "Alveolar macrophages from persons living with HIV show impaired epigenetic response to Mycobacterium tuberculosis." *The Journal of clinical investigation* vol. 131,22 (2021): e148013. <doi:10.1172/JCI148013>

Evans, C., Hardin, J., Stoebel, D. M. "Selecting between-sample RNA-Seq normalization methods from the perspective of their assumptions." *Briefings in Bioinformatics*. (2018).DOI: 10.1093/bib/bbx008

Isserlin, R." Lecture 4 - Exploring the data and basics of Normalization" (2021).

Isserlin, R. "Lecture 5 - Data exploration and Identifier mapping" (2022).

Kempenaar, M. "Analysis of Gene Expression" (2021). <https://mkempenaar.github.io/gene_expression_analysis/>

Love M., Anders S., Huber W. "RNA-Seq workflow: gene-level exploratory analysis and differential expression" (2014). <https://www.bioconductor.org/help/course-materials/2014/SeattleOct2014/B02.1.1_RNASeqLab.html>
