---
title: "Assignment 2"
author: "Casey Hon"
date: "r Sys.Date()"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 2
---
```{r, message = FALSE}
suppressWarnings({
  if (! requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
  }
  if (! requireNamespace("Biobase", quietly = TRUE)) {
    BiocManager::install("Biobase")
  }
  if (! requireNamespace("limma", quietly = TRUE)) {
    BiocManager::install("limma")
  }
  if (! requireNamespace("knitr", quietly = TRUE)) {
    BiocManager::install("knitr")
  }
  if (! requireNamespace("ggrepel", quietly = TRUE)) {
    BiocManager::install("ggrepel")
  }
})
```

## Introduction
In Assignment #1, I analyzed RNA-seq data from the study *Alveolar macrophages from persons living with HIV show impaired epigenetic response to Mycobacterium tuberculosis* (Correa-Macedo et al., 2021). The data contained 16 control subjects who were HIV-free (HC), 20 persons living with HIV receiving ART (PLWH), and 14 subjects who received ART as preexposure prophylaxis (PrEP) to prevent HIV infection. Each sample was challenged with *M. tuberculosis* in vitro.

Some basic statistics:
GEO ID: GSE165708

The original dataset had over 60,000 genes. After using edgeR filtering protocols, 13,780 genes were left. 8 genes were duplicates so those were removed. The data was then normalized using the Trimed Mean of M-values (TMM) method. After normalizing, a Multi-Dimensional Scaling (MDS) plot was used to visualize clustering, but no clear clustering by sample group was seen. 

```{r, message = FALSE}
library(ggplot2)
mds <- readRDS("mds.rds")
qplot(X1, X2, color = sample_group, data = mds, main = "MDS plot of Poisson Distances by sample group")
```

Thus, this notebook aims to take this normalized expression data and:
- rank genes according to differential expression;
- perform thresholded over-representation analysis (ORA);
- and highlight dominant themes in the top set of genes.

## Differential Gene Expression
From the first assignment, we know that the possible groups were sample group, patient, and whether or not the sample was challenged. Based on the MDS plot of the normalized data, it seems like there is a pattern that the Healthy Control (HC) group (in pink) tends to be on the left side, the Persons living with HIV (PLWH) group (in green) tending to be more on the right, and the pre-exposure prophylaxis (PrEP) group (in blue) towards the bottom middle. Although this wasn't a very clear distinction of clusters, but plotting by other conditions, by patient or by challenged status made the clustering  even more obscure. Thus, I'm choosing to design my model by sample group (HC, PLWH, and PrEP).

```{r}
library(Biobase)
library(limma)
library(knitr)
# create design matrix
groups <- readRDS("groups.rds")
cond_mod <- model.matrix(~ groups$sample_group)

# create data matrix
normed_log_counts <- readRDS("normed_log_counts.rds")
rownames(normed_log_counts) <- make.unique(rownames(normed_log_counts)) # TO-DO: why are there duplicated rownames...? how is that possible?
es <- ExpressionSet(assayData = normed_log_counts)

# test for differential expression
fit <- lmFit(es, design = cond_mod)
fit2 <- eBayes(fit, trend = TRUE) # calculates p-values too

# sort by p-value
pval_thres <- 0.05 
fit_pvals <- data.frame(fit2$p.value)
colnames(fit_pvals) <- c("pvalue", "sample_groupPLWH", "sample_groupPrEP")
fit_pvals <- fit_pvals[order(fit_pvals$pvalue), ]
fit_pvals$pvalue <- as.character(fit_pvals$pvalue)
kable(fit_pvals[1:10, ], row.names = TRUE, format = "markdown",
              caption = "Table 1: Top 10 genes with lowest p-values")
```
We can see from table 1 that just the top 10 genes with smallest p-values have nearly 0 as p-values, which means the null hypothesis is rejected (there is strong differential expression). 

There are exactly `r length(which(fit_pvals$pvalue < pval_thres))` genes that have a p-value of less than 0.05, which is `r length(which(fit_pvals$pvalue < pval_thres)) / nrow(fit_pvals) * 100`%. 

However, it's important to note these statistics are before correcting p-values for multiple testing. We'll adjust for this using the Benjamini-Hochberg Procedure, which controls for the fact that some small p-values may occur by chance (Glen, n.d.).

```{r}
# correct p-values with mult hyp correction, sorted by adjusted p-value
mhc <- topTable(fit2, coef = ncol(cond_mod), adjust.method = "BH", 
                number = nrow(es), sort.by = "p")
knitr::kable(mhc[1:10, c("P.Value", "adj.P.Val")], row.names = TRUE, format = "markdown", digits = 32, caption = "Table 2: Top 10 genes with lowest adjusted p-values")
```
Based on table 2, we see that the 10 genes with the lowest p-values are completely different. Now, there are `r length(which(mhc$adj.P.Val < pval_thres))` genes that have a p-value of less than 0.05, which is `r length(which(mhc$adj.P.Val < pval_thres)) / nrow(mhc) * 100`%. 

```{r}
# referred to https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html
# volcano plot, highlight genes of interest
log_thres <- 1.5 

mhc$diffexpressed <- "none"
mhc$diffexpressed[mhc$logFC > log_thres & mhc$P.Value < pval_thres] <- "up" # up-regulated
mhc$diffexpressed[mhc$logFC < -log_thres & mhc$P.Value < pval_thres] <- "down" # down-regulated

mhc$delabel <- NA
mhc$gene_name <- rownames(mhc)
mhc$delabel[mhc$diffexpressed != "none"] <- mhc$gene_name[mhc$diffexpressed != "none"]

volc <- ggplot(data = mhc, aes(x=logFC, y=-log10(P.Value),
               col = diffexpressed, label = delabel)) +
  geom_point() +
  theme_minimal() +
  ggrepel::geom_text_repel() +
  labs(title = "Volcano Plot highlighting Differentially-Expressed Genes",
       x = "log fold change", y = "-log10(p-value)", 
       col = "Differential Expression") +
  scale_color_manual(values = c("blue", "black", "red"), labels = c("down-regulated", "none", "up-regulated")) +
  geom_hline(yintercept = -log10(pval_thres), col = "red") +
  geom_vline(xintercept = c(-log_thres, log_thres), col = "red") 
volc

```

Out of `r nrow(mhc)` genes total, `r length(mhc$gene_name[mhc$diffexpressed == "up"])` genes were up-regulated, and `r length(mhc$gene_name[mhc$diffexpressed == "down"])` genes were down-regulated.

```{r}
kable(mhc[mhc$diffexpressed == "up", c("logFC", "P.Value", "adj.P.Val")],
      caption = "Up-regulated genes",
      col.names = c("log fold change", "p-value", "adjusted p-value"))
kable(mhc[mhc$diffexpressed == "down", c("logFC", "P.Value", "adj.P.Val")],
      caption = "Down-regulated genes",
      col.names = c("log fold change", "p-value", "adjusted p-value"))
```

```{r}
# visualize top hits with heatmap
library(pheatmap)

hm_matrix <- normed_log_counts[which(mhc$diffexpressed == "up" | mhc$diffexpressed == "down"),]
hm_matrix[which(!is.finite(hm_matrix))] <- 0 # remove infinite value
hm_matrix <- t(scale(t(hm_matrix))) # scale to normalize
pheatmap(hm_matrix, show_colnames = FALSE)

```

1. Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?

There are exactly `r length(which(fit_pvals$pvalue < pval_thres))` genes that have a p-value of less than 0.05, which is `r length(which(fit_pvals$pvalue < pval_thres)) / nrow(fit_pvals) * 100`%. 
I chose to stay with the conventional p-value threshold of 0.05 based on arguments by Di Leo & Sardanelli (2020). Achieving enough power with a lower threshold for p-value would require larger sample sizes, which can be difficult especially with this line of research and acquiring the type of data that this study uses (Di Leo & Sardanelli, 2020).
I used a 1.5 fold change threshold to determine whether genes were up or down regulated, based on the opinions of a couple forum posts (Sulaiman, 2018). 

2. Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?

I chose to use the default Benjamini-Hochberg (BH) method for correction as it is recommended as one of the most superior in terms of user-friendliness and documentation, and since it is the default in limma too so using this method could increase probability of potential comparison with other analyses (Korthauer et al., 2019; Ritchie et al., 2015). IHW was also ranked as a top method, but I chose against this since it demonstrated low False Discovery Rate (FDR) compared to other methods (Korthauer et al., 2019). However, IHW did have very similar patterns to BH in terms of FDR control, applicability, and usability (Korthauer et al., 2019). `r length(which(mhc$adj.P.Val < pval_thres))` genes passed correction.

3. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.

4. Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not.


## Thresholded Over-Representation Analysis
1. Which method did you choose and why?

2. What annotation data did you use and why? What version of the annotation are you using?

3. How many genesets were returned with what thresholds?

4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?


## Interpretation
1. Do the over-representation results support conclusions or mechanism discussed in the original paper?

2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.


## References

Correa-Macedo, W., et al. "Alveolar macrophages from persons living with HIV show impaired epigenetic response to Mycobacterium tuberculosis." *The Journal of clinical investigation* vol. 131,22 (2021): e148013. <doi:10.1172/JCI148013>

Di Leo, G., Sardanelli, F. Statistical significance: p value, 0.05 threshold, and applications to radiomics—reasons for a conservative approach. Eur Radiol Exp 4, 18 (2020). https://doi.org/10.1186/s41747-020-0145-y

Glen, S. "Benjamini-Hochberg Procedure" From StatisticsHowTo.com: Elementary Statistics for the rest of us! https://www.statisticshowto.com/benjamini-hochberg-procedure/

Isserlin, R. "Lecture 6 - Differential Expression" (2022).

Korthauer, K., Kimes, P.K., Duvallet, C. et al. A practical guide to methods controlling false discoveries in computational biology. Genome Biol 20, 118 (2019). https://doi.org/10.1186/s13059-019-1716-1

Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15:550. 10.1186/s13059-014-0550-8

Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015). “limma powers differential expression analyses for RNA-sequencing and microarray studies.” Nucleic Acids Research, 43(7), e47. doi: 10.1093/nar/gkv007.

Sulaiman, JE. (2017). "Do you know what is the threshold for up and down-regulation in proteomics?" ResearchGate. https://www.researchgate.net/post/Do-you-know-what-is-the-threshold-for-up-and-down-regulation-in-proteomics
